% !TEX TS-program = pdflatex
% !TEX encoding = UTF-8 Unicode

% This is a simple template for a LaTeX document using the "article" class.
% See "book", "report", "letter" for other types of document.

\documentclass[11pt,hyperref,draft]{article} % use larger type; default would be 10pt


%%% Examples of Article customizations
% These packages are optional, depending whether you want the features they provide.
% See the LaTeX Companion or other references for full information.

%%% PAGE DIMENSIONS
\usepackage{xcolor} % to change the page dimensions
\usepackage{geometry} % to change the page dimensions
\geometry{a4paper} % or letterpaper (US) or a5paper or....
% \geometry{margin=2in} % for example, change the margins to 2 inches all round
% \geometry{landscape} % set up the page for landscape
%   read geometry.pdf for detailed page layout information

\usepackage{graphicx} % support the \includegraphics command and options

% \usepackage[parfill]{parskip} % Activate to begin paragraphs with an empty line rather than an indent

%%% PACKAGES
\usepackage{amsmath} % for much better looking tables
\usepackage{xspace}
\usepackage{url}
\usepackage{amssymb} % for much better looking tables
\usepackage{booktabs} % for much better looking tables
\usepackage{array} % for better arrays (eg matrices) in maths
\usepackage{ntheorem} % for better arrays (eg matrices) in maths
\usepackage{paralist} % very flexible & customisable lists (eg. enumerate/itemize, etc.)
\usepackage{verbatim} % adds environment for commenting out blocks of text & for better verbatim
\usepackage{subfig} % make it possible to include more than one captioned figure/table in a single float
\usepackage{framed}
% These packages are all incorporated in the memoir class to one degree or another...

%%% HEADERS & FOOTERS
\usepackage{fancyhdr} % This should be set AFTER setting up the page geometry
\pagestyle{fancy} % options: empty , plain , fancy
\renewcommand{\headrulewidth}{0pt} % customise the layout...
\lhead{}\chead{}\rhead{}
\lfoot{}\cfoot{\thepage}\rfoot{}

%%% SECTION TITLE APPEARANCE
\usepackage{sectsty}
\allsectionsfont{\sffamily\mdseries\upshape} % (See the fntguide.pdf for font help)
% (This matches ConTeXt defaults)

%%% ToC (table of contents) APPEARANCE
\usepackage[nottoc,notlof,notlot]{tocbibind} % Put the bibliography in the ToC
\usepackage[titles,subfigure]{tocloft} % Alter the style of the Table of Contents
\renewcommand{\cftsecfont}{\rmfamily\mdseries\upshape}
\renewcommand{\cftsecpagefont}{\rmfamily\mdseries\upshape} % No bold!

\newcommand{\Answer}[1]{\noindent\textsf{\textbf{Response: }}{\sf#1}\\}
\newcommand{\Comment}[1]{\noindent\textsf{\textbf{Comment: }}{\it#1}\\[.5em]}

%%% END Article customizations

\input{macros}

%%% The "real" document content comes below...

\title{Using structural and evolutionary information to detect and correct pyrosequencing errors in non-coding RNAs}
\author{Vladimir Reinharz$^1$, Yann Ponty$^{2,*}$, J\'er\^{o}me Waldisp\"{u}hl$^{1,*}$}
\date{$^1$~School of Computer Science, McGill University, Montreal, Canada, $^2$~Laboratoire d'informatique, \'Ecole Polytechnique, Palaiseau, France.\\ \small $^*$~Corresponding authors: \texttt{jeromew@cs.mcgill.ca}, \texttt{yann.ponty@lix.polytechnique.fr}}

%\date{} % Activate to display a given date or no date (if empty),
         % otherwise the current date is printed 

\begin{document}
\maketitle

In general, the reviewers did not raise very specific points, but were mostly concerned by the applicability of the method in a high-throughput context. In this extended version, we demonstrated the scalability of the method by running processing longer sequences (16S rRNAs) obtained from random reads (generated by the {\tt ART} pyrosequencing simulator), a setting which is closer to real use-cases in metagenomics. Furthermore, our python implementation  is still more time-consuming than would be ideal in a high-throughput context. However, we propose a pipeline which first maps short reads to the reference alignment, followed by an alignment/correction of the consensus sequence against the reference multiple sequence alignment. In this setting, {\tt RNAPyo} is only executed once for each rRNA, as opposed to once for every read, and the typical runtime of our software does not induce such a strong overhead. Additionally, we investigated the error-correction potential as a function of the coverage (depth) of the sequencing.

\section*{Detailed responses}

\Comment{(Reviewer \#1) While the paper is algorithmic interesting, and clearly publishable in principle, I'm a bit concerned with the practical application of RNApyro. If many reads are available for the same sequence, the tool is obviously not necessary since purely sequence-based error detection should be applicable. In the scenario envisioned here, however, the model makes a very strong assumption, name a completely known consensus structure to which the query sequence must fit without indels. While plausible if the consensus is derived from closely related species (in which case point mutations that disrupt the structure are easily visible in most cases) one would expect also structural variations as well.\\
(Reviewer \#2) However, assuming that it works as expected, I understand that a successful "error correction" involves replacing an unlikely secondary structure base pair by a more likely one. Couldn't this become a problem for outlier RNA sequences that do not exactly fold according to the canonical model used in RFAM? If this is the case, then the program would correct certain sequences/species better than others, which is not quite a desirable property in error correction. 
}
\Answer{We agree with the reviewer that the error-correction potential of our method may be challenged by sequences whose folding is substantially different from the consensus structure. However, assuming some level of structural conservation within RNA families that share a given function is generally held as reasonable, and constitutes the main paradigm in RNA comparative bioinformatics, e.g. underlying the RFAM database and homologous sequences retrieval. This paradigm is, of course, challenged in the case of structure-disrupting mutations (which are increasingly identified and linked with RNA-based genetic conditions), but should remain reasonably consensual in the context of metagenomics, where structural outliers (often pathological, or even lethal) are perhaps of lesser interest. \\
Furthermore, the consensus structure needs not be completely known. In this extended version, we considered the potential for error-correction of {\tt RNAPyro}, restricted paired regions, based on the dual rationale that 1) our energy model does not (yet!) capture unpaired regions, and 2) unpaired region may be paired but not sufficiently conserved throughout the evolution (or simply misaligned). The excellent areas under the ROC curves observed in this case suggests that a partial consensus may not too strongly impede the quality of our error corrections. \\
Finally, regarding a more general structural diversity in families of structural RNAs (e.g. only preserving a domain, a opposed to a general architecture), we believe such cases would be revealed by the level of detail of the consensus structure and, by disallowing error-corrections within unpaired regions, our method would not overly erroneous and inhomogeneous. }

\Comment{(Reviewer \#1) Also, how realistic is it that one would trust a single 454 read with 6 or 12 expected errors ? In practice, on would at least want to have a prior of the quality of the nucleotide call included.}
\Answer{We addressed this comment by estimating the number of errors from the coverage. We simply fitted the number of errors empirically observed on sequences assembled from Illumina simulated short reads, as a function of the coverage and sequence length.}



\Comment{(Reviewer \#2) Another issue is that of computing time. Although speed may be improved relative to prior versions of the algorithm, the times presented in Table I (is it for a single sequence?) do not appear to be compatible with NGS applications.}
\Answer{We improved the speed of our implementation by performing further preprocessing, and show in the extended version a more complete time benchmark. As noted earlier in this response, we run {\tt RNApyro} on a consensus sequence obtained by mapping sequences onto the multiple sequence alignment. Consequently, one does not need to run the software once for every read, but rather once every (Length$\times$Coverage$/$Read-Length)-th read, and the improve runtime should now realistically allow for embedding {\tt RNAPyro} within a high-throughput pipeline.
}



\Comment{(Reviewer \#3) The problem setting is unique, but applicability may be limited to known (annotated) ncRNAs. The algorithm may be more naturally applied to optimization by computational evolution.}
\Answer{As suggested by the reviewer, we leveraged the inside computation into a 	statistical sampling algorithm addressing the inverse folding (aka RNA design) problem (manuscript accepted, scheduled for an oral presentation at the ISMB'13 conference~\cite{Reinharz2013}).}

\Comment{(Reviewer \#2) Typos: Please check sentences...}
\Answer{Fixed}

\bibliographystyle{amsplain}
\bibliography{RNApyro}
\end{document}
