%!TEX root = main_RECOMB.tex
\section{Methods}
\label{sec:methods}
Let $\Omega$ be an un-gapped RNA alignment, $S$ its associate secondary structure, $s$ an RNA 
sequence and $m\geq 0$. $S$ is considered as  one derivation of the SCFG generating  
all possible secondary structures of the same length as $s$. We define a variant of the
 \emph{Inside-Outside algorithm}~\cite{Lari1990} with the two following functions,
$\Z{i,j}{m}{a,b}$ and $\Y{i,j}{m}{a,b}$. The former, ass the 
equivalent of the \emph{inside}, computes the partition function between $i$ and $j$ included, 
knowing that position $i-1$ is composed of nucleotide $a$ (resp. $j+1$ is $b$) and containing $m$ mutations. The latter computes the \emph{outside}, in particular, knowing 
that position $i+1$ is composed of $a$ (resp. $j-1$ is $b$) and containing, outside 
including $i$ and $j$, $m$ mutations.
The Boltzmann weights are a combination of the base pairs stacking energy, 
using as values those of the NNDB~\cite{Turner2010}, and
the average isostericity with $\Omega$, as defined in~\cite{Stombaugh2009}. 

This allows us to compute the probability of any position being a specific nucleotides,
over all sequences having at most $m$ mutations from $s$, under the SCFG derivation $S$.

\subsection{Definitions}
Let be $B:=\left\{\text{A},\text{C},\text{G},\text{U}\right\}$, the set of nucleotides.
Given $s\in B^n$ an RNA sequence, let $s_i$ be the nucleotide at position $i$. Let $\Omega$ be a set of un-gapped RNA sequences of
length $n$, and $S$ a secondary structure without pseudoknots. 
Formally, if $(i,j)$ and $(k,l)$ are base pairs in $S$, there is no overlapping extremities
 $\{i,j\}\cap \{k,l\}=\varnothing$ and either the intersection is empty 
 ($[i,j]\cap[k,l]=\varnothing$) or one is included in the other ($[k,l]\subset[i,j]$ or 
 $[i,j]\subset[k,l])$. Let $R$ be the Boltzmann constant, $T$ the temperature in Kelvin and
  the function $\delta$ such that: 
 $\forall a,a' \in B, \delta_{a,a'}:=\left\{\begin{array}{ll}
															1 & \text{If } a\equiv a'\\
															0 & \text{Else}
														\end{array}\right.$

\subsection{Energy Model}
The energy we will be composed of two function, $ES^{\beta}_{ab\to a'b'}$ and 
$EI_{ab\to a'b'}$. $ES^\beta_{ab\to a'b'}$ is equal to the 
stacking energy of the base pair with nucleotides $ab$ on top of the base pair with nucleotides 
$a'b'$, as set in the NNDB~\cite{Turner2010}. If one of the base pair is not valid (i.e. not in 
$\{\text{GU},\text{UG},\text{CG},\text{GC}, \text{AU or UA}\}$, the value is a parameter 
$\beta \in [1,\infty]$. This allows
to completely forbid a sequence where a base pair is non valid, when $\beta = \infty$ or only 
penalize it.
$EI_{ab\to a'b'}$ is the isostericity between
 the base pair $ab$ and $a'b'$ as defined in~\cite{Stombaugh2009}. Let be $\alpha\in[0,1]$, it 
 will be used to balance the weight given to the stacking energy or the isostericity.
	
\subsection{Inside}
We define the \emph{Inside} function $\Z{i,j}{m}{a,b}$ as follows.
\subsubsection{Initial conditions}
$$
	\forall i \in (0,\cdots,n-1):\, \Z{i+1,i}{m}{a,b}=\left\{
	\begin{array}{ll}
		1 &\text{If } m = 0\\
		0 &\text{Else }
	\end{array}\right.
$$
\subsubsection{Recursion}
$$
	\Z{i,j}{m}{a,b}:=\left\{
  \begin{array}{ll}
  		\displaystyle
      \sum_{\substack{a'\in \B,\\ \Kron_{a',s[i]}\le m}}  
      \Z{i+1,j}{m-\Kron_{a',s[i]}}{a',b} & \text{If }S{[i]}=-1\\
      \displaystyle
      \sum_{\substack{a',b'\in \B^2,\\ \Kron_{a'b',s[i]s[j]}\le m}}  
			 e^{\frac{-ES_{a b \to a' b'}}{RT}}
			 \Z{i+1,j-1}{m-\Kron_{a'b',s[i]s[j]}}{a',b'}&
			 \text{Elif }S{[i]}=j \land S{[i-1]}=j+1\\
			 \displaystyle
      \sum_{\substack{a',b'\in \B^2,\\ \Kron_{a'b',s[i]s[k]}\le m}}
      \sum_{m'=0}^{m-\Kron_{a'b',s[i]s[k]}}
   		 e^{\frac{-EI_{s[i]s[k],a'b'}}{RT}}
      \Z{i+1,k-1}{m-\Kron_{a'b',s[i]s[k]}-m'}{a',b'}
      \Z{k+1,j}{m'}{b',b} & \text{Elif }S[i]=k \land i < k \leq j\\
      0 &\text{Else}
	\end{array}\right.
$$

\subsection{Outside}	
\subsubsection{Initial conditions}
$$
	\Y{-1,j}{m}{X,X}:=
		\displaystyle
	  \Z{j,n-1}{m}{X,X}
$$
\subsubsection{Recursion}
$$
	\Y{i,j}{m}{a,b} = \left\{
  \begin{array}{ll}
		\displaystyle
    \sum_{\substack{a'\in \B,\\ \Kron_{a',s[i]}\le m}}
    \Y{i-1,j}{m- \Kron_{a',s[i]}}{a',b} &
    \text{Elif }S[i]=-1 \\
    \displaystyle
    \sum_{\substack{a'b'\in \B^2,\\ \Kron_{a'b',s[i]s[j]}\le m}}
		 e^{\frac{-ES_{ a' b' \to a b }}{RT}}
    \Y{i-1,j+1}{m- \Kron_{a'b',s[i]s[j]}}{a',b'} &
   	 \text{Elif }S{[i]}=j \land S{[i+1]}=j-1\\
		 \displaystyle
		 \sum_{\substack{a'b'\in \B^2,\\ \Kron_{a'b',s[i]s[k]}\le m}}
		 \sum_{m'=0}^{m-\Kron_{a'b',s[i]s[k]}}
  		 e^{\frac{-EI_{s[i]s[k],a'b'}}{RT}}
		 \Y{i-1,k+1}{m- \Kron_{a'b',s[i]s[k]} - m'}{a',b'}
     \Z{j,k-1}{m'}{b,b'} &
		 \text{Elif }S{[i]}=k \geq j\\
		 \displaystyle
		 \sum_{\substack{a'b'\in \B^2,\\ \Kron_{a'b',s[k]s[i]}\le m}}
		 \sum_{m'=0}^{m-\Kron_{a'b',s[k]s[i]}}
  		 e^{\frac{-EI_{s[i]s[k],a'b'}}{RT}}
		 \Y{k-1,j}{m- \Kron_{a'b',s[k]s[i]} - m'}{a',b}
     \Z{k+1,i-1}{m'}{a',b'} &
		 \text{Elif }-1 < S{[i]}=k < i\\
		 0 & \text{Else}
  \end{array}\right.
$$
\subsection*{}
We must have (if $S[k] = -1)$:
$$
	\Z{0,n-1}{m}{X,X} \equiv     
	\sum_{\substack{a\in \B,\\ \Kron_{a,s[k]}\le m}}	
	\Y{k-1,k+1}{m-\Kron_{a,s[k]}}{a,a};\qquad
	\forall k \in \{0,\cdots,n-1\}
$$

\subsection{Probability of a position being a given nucleotide}
Given $x\in B$,
$$
	\mathbb{P}(s[i] = x\mid s, S,m):=\left\{
	\begin{array}{ll}
		\displaystyle
		\frac{
			\Y{i-1,i+1}{m-\Kron_{x,s[k]}}{x,x}
		}{
			\Z{0,n-1}{m}{X,X}
		}
		&\text{If }S[i] = -1\\
		\displaystyle
 		\frac{
			\displaystyle
			\sum_{\substack{b\in Bases\\\Kron_{bx,s[k]s[i]\leq m}}}
			\sum_{m'=0}^{m-\Kron_{bx,s[k]s[i]}}
			\Y{k-1,i+1}{m-\Kron_{bx,s[k]s[i]-m'}}{b,x}
			\Z{k+1,i-1}{m'}{b,x}
		}{
			\Z{0,n-1}{m}{X,X}
		}
		&\text{If }S[i]=k<i\\
		\displaystyle
		\frac{
			\displaystyle
			\sum_{\substack{b\in Bases\\\Kron_{xb,s[i]s[k]\leq m}}}
			\sum_{m'=0}^{m-\Kron_{xb,s[i]s[k]}}
			\Y{i-1,k+1}{m-\Kron_{xb,s[i]s[k]-m'}}{x,b}
			\Z{i+1,k-1}{m'}{x,b}
		}{
			\Z{0,n-1}{m}{X,X}
		}
		&\text{If }S[i]=k>i
	\end{array}\right.
$$
